src/circuit/gadgets/constraints.rs

```diff
@@ -427,6 +464,148 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
     }
 }
 
+// Enforce div and mod operation for U64. We need to show that
+// arg1 = q * arg2 + r, such that 0 <= r < arg2.
+pub fn enforce_u64_div_mod<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    cond: Boolean,
+    arg1: &AllocatedPtr<F>,
+    arg2: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>), SynthesisError> {
+    let arg1_u64 = match arg1.hash().get_value() {
+        Some(v) => v.to_u64_unchecked(),
+        None => 0, // Blank and Dummy
+    };
+    let arg2_u64 = match arg2.hash().get_value() {
+        Some(v) => v.to_u64_unchecked(),
+        None => 0, // Blank and Dummy
+    };
+
+    let (q, r) = if arg2_u64 != 0 {
+        (arg1_u64 / arg2_u64, arg1_u64 % arg2_u64)
+    } else {
+        (0, 0)
```

Comment:
Maybe this won't matter in the end (still reading below), but I note that this will assign 0 to `r`, when it's an undefined situation. Currently, evaluation panics in this situation, but it should be an explicitly checked error. (#219) We need to make sure that the circuit does the right thing in that case.

In particular, it would be an error if you could produce a proof that `(% n 0)` => `0`.

Reply:
Got catch! I included the error in the circuit, and a unit test for it.